### PR TITLE
added slider and popup

### PR DESCRIPTION
**Pull request by sahandfarahani**
Slider

### CODE REVIEW

File: src/Inputs/Slider.tsx [hunk 1]
```diff
@@ -1,79 +1,440 @@
...
-        height: 6px;
-        cursor: pointer;
-        animate: 0.2s;
-        box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
-        background: #3071a9;
-        border-radius: 10px;
-        border: 0.2px solid ${(props): string => props.trackColor || '#010101'};
+}): React.ReactElement => {
+    // DOM Elements
+    const bar = useRef() as RefObject<HTMLDivElement>;
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:65]: const bar= useRef<HTMLDivElement>();
Can you do this instead?

File: src/Inputs/Slider.tsx [hunk 2]
```diff
@@ -1,79 +1,440 @@
...
+    const finishThumb = useRef() as RefObject<HTMLDivElement>;
+    const marksBar = useRef() as RefObject<HTMLDivElement>;
+
+    // Thumb Positions in Px
+    const [finishThumbLeft, setFinishThumbLeft] = useState(max);
+    const [startThumbLeft, setStartThumbLeft] = useState(min);
+
+    const MaxAndMinDifference = max - min;
+
+    let draggingThumb = 'Finish';
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:77]: Why are these things being computed on every render?

File: src/Inputs/Slider.tsx [hunk 3]
```diff
@@ -1,79 +1,440 @@
...
+
+    // Translate a value to Pixel
+    const translateToPixels = (theValue: number): number => {
+        const pixelTranslator =
+            (bar.current?.clientWidth as number) / MaxAndMinDifference;
+        return ((theValue as number) - min) * pixelTranslator;
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:83]: do you need as Number if the type is number??

File: src/Inputs/Slider.tsx [hunk 4]
```diff
@@ -1,79 +1,440 @@
...
+    function resizing(): void {
+        if (finishThumbLeftInValue) {
+            setFinishThumbLeft(translateToPixels(finishThumbLeftInValue));
+        }
+        if (startThumbLeftInValue) {
+            setStartThumbLeft(translateToPixels(startThumbLeftInValue));
+        }
+
+        if (marksBar.current && marks) {
+            const theMarks = marksBar.current.children as any;
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:160]: children are ReactNode ?

File: src/Inputs/Slider.tsx [hunk 5]
```diff
@@ -1,79 +1,440 @@
...
-        width: 16px;
-        border-radius: 50%;
-        background: ${(props): string => props.buttonColor || '#ffffff'};
-        cursor: pointer;
-        -webkit-appearance: none;
-        margin-top: -5px;
+
+    window.onresize = resizing;
+
+    function onMouseMove(theevent: MouseEvent): void {
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:176]: Does ES6 notation not work for this?

File: src/Inputs/Slider.tsx [hunk 6]
```diff
@@ -1,79 +1,440 @@
...
+                    </SliderThumbStart>
+                )}
+
+                <SliderThumb
+                    ref={finishThumb}
+                    disabled={disabled}
+                    hasTwoInputs={hasTwoInputs}
+                    left={finishThumbLeft}
+                    onMouseDown={(event): any =>
+                        handleMouseDown(event as MouseEvent)
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:281]: Can't you just type event: MouseEvent?

File: src/Inputs/Slider.tsx [hunk 7]
```diff
@@ -1,79 +1,440 @@
...
+            </SliderBoard>
+            {marks && (
+                <SliderBoardMarks
+                    ref={marksBar}
+                    theme={theme}
+                    disabled={disabled}
+                >
+                    {marks.map(
+                        ({ key, mark }): React.ReactElement => (
+                            <div
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:304]: Where is the key?

File: src/Inputs/Slider.tsx (hunk 1 above)
Comment (sahandfarahani) [reply src/Inputs/Slider.tsx:65]: I'll get typescript errors whenever im using the props of bar

Comment (ralph-dev) [reply src/Inputs/Slider.tsx:65]: I mean
useRef< HTMLDivElement >()

File: src/Inputs/Slider.tsx (hunk 4 above)
Comment (sahandfarahani) [reply src/Inputs/Slider.tsx:160]: When I put reactnode I get the ts error that reactnode is not assignable to htmlelement... If I change child type I get error that child.styles does not exist on string...

File: src/Inputs/Slider.tsx (hunk 1 above)
Comment (sahandfarahani) [reply src/Inputs/Slider.tsx:65]: I'll get ts errors that mutablerefobject<htmldivelement/undefined> is not assignable to type reconnect<htmldivelement> and ((instance:htmldivelement/null)=>voidect.

File: src/Inputs/Slider.tsx (hunk 7 above)
Comment (sahandfarahani) [reply src/Inputs/Slider.tsx:304]: Sorry I don't get this

Comment (ralph-dev) [reply src/Inputs/Slider.tsx:304]: https://reactjs.org/docs/lists-and-keys.html

File: src/Inputs/Slider.tsx (hunk 2 above)
Comment (sahandfarahani) [reply src/Inputs/Slider.tsx]: put I put these in Useeffect or UseLayouteffect they will get undefined in the body and it's like they don't exist

File: src/Inputs/Slider.tsx [hunk 8]
```diff
@@ -1,79 +1,436 @@
...
+    const selectedBar = useRef() as RefObject<HTMLDivElement>;
+    const startThumb = useRef() as RefObject<HTMLDivElement>;
+    const finishThumb = useRef() as RefObject<HTMLDivElement>;
+    const marksBar = useRef() as RefObject<HTMLDivElement>;
+
+    // Thumb Positions in Px
+    const [finishThumbLeft, setFinishThumbLeft] = useState(max);
+    const [startThumbLeft, setStartThumbLeft] = useState(min);
+
+    const MaxAndMinDifference = max - min;
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:75]: Fix

File: src/Inputs/Slider.tsx [hunk 9]
```diff
@@ -10,6 +10,8 @@ import styled, { DefaultTheme } from 'styled-components';
 import { LabelLayout, LabelLayoutProps } from '@Layouts';
 import { Popup } from '../Containers/Popup';
 
+let draggingThumb = 'Finish';
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:13]: still incorrect.

File: src/Inputs/Slider.tsx [hunk 10]
```diff
@@ -59,6 +62,7 @@ export const Slider: React.FunctionComponent<SliderProps> = ({
     popupwidth = 'auto',
     popupheight = 20,
     theme,
+    maxAndMinDifference = max - min,
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:65]: Doesnt this only change when min or max changes? if so place in useeffect and watch the two variables

File: src/Inputs/Slider.tsx (hunk 9 above)
Comment (sahandfarahani) [reply src/Inputs/Slider.tsx:13]: I can use usestate for this then

Comment (sahandfarahani) [reply src/Inputs/Slider.tsx:13]: But why would it be incorrect now that it does not update with every render?

File: src/Inputs/Slider.tsx (hunk 10 above)
Comment (sahandfarahani) [reply src/Inputs/Slider.tsx:65]: The thing is it never changes because min and max never change... The only reason Im using it is to make the codes shorter and more readable ... I can delete it and replace it everywhere with max-min simply

Comment (ralph-dev) [reply src/Inputs/Slider.tsx:65]: @sahandfarahani Yes do that then

File: src/Inputs/Slider.tsx (hunk 9 above)
Comment (ralph-dev) [reply src/Inputs/Slider.tsx:13]: Let's jump on a call afterwards.

File: src/Inputs/Slider.tsx (hunk 1 above)
Comment (fpunny) [reply src/Inputs/Slider.tsx:65]: if you read the error message it says instance has to be null or the htmldivelement.

File: src/Inputs/Slider.tsx [hunk 11]
```diff
@@ -1,79 +1,456 @@
...
+    theme?: DefaultTheme;
 }
 
 export const Slider: React.FunctionComponent<SliderProps> = ({
+    disabled = false,
+    marks,
+    min = 0,
+    max = 100,
+    hasRail,
+    hasTwoInputs = false,
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:51]: @sahandfarahani Is this hasTwoInputs supposed to be hasTwoKnobs? What is this?

File: src/Inputs/Slider.tsx [hunk 12]
```diff
@@ -1,79 +1,456 @@
...
+    valueFinish = max,
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:52]: @sahandfarahani What does this variable mean?

File: src/Inputs/Slider.tsx (hunk 1 above)
Comment (indancisive) [reply src/Inputs/Slider.tsx:65]: you have to define the ref as null as the ref must either be the element object or null, not undefined
image

File: src/Inputs/Slider.tsx [hunk 13]
```diff
@@ -1,79 +1,456 @@
...
-        background: #367ebd;
-    }
-    &::-moz-range-track {
+}): React.ReactElement => {
+    // DOM Elements
+    const bar = useRef() as RefObject<HTMLDivElement>;
+    const selectedBar = useRef() as RefObject<HTMLDivElement>;
+    const leftThumb = useRef() as RefObject<HTMLDivElement>;
+    const rightThumb = useRef() as RefObject<HTMLDivElement>;
+    const marksBar = useRef() as RefObject<HTMLDivElement>;
```
Comment (indancisive) [inline src/Inputs/Slider.tsx:69]: it would be easier to read the later code if you name these marksBarRef, etc. just as a convention

File: src/Inputs/Slider.tsx (hunk 12 above)
Comment (ryanarine) [reply src/Inputs/Slider.tsx:52]: I think that's supposed to be one of the initial values (valueStart being the other)

File: src/Inputs/Slider.tsx [hunk 14]
```diff
@@ -1,79 +1,456 @@
...
+
+    // Final Result
+    useMemo((): void => {
+        onChange({
+            values: {
+                lowValue: leftThumbPositionInValue,
+                highValue: rightThumbPositionInValue,
+            },
+        });
+    }, [rightThumbPositionInValue, leftThumbPositionInValue]);
```
Comment (indancisive) [inline src/Inputs/Slider.tsx:128]: don't use useMemo for sideeffects

File: src/Inputs/Slider.tsx [hunk 15]
```diff
@@ -1,79 +1,456 @@
...
+import styled, { DefaultTheme } from 'styled-components';
+import { LabelLayout, LabelLayoutProps } from '@Layouts';
+import { Popup } from '../Containers/Popup';
+
+export interface MarkProps {
+    key: number;
+    mark: string;
+}
+
+export interface ResultValues {
```
Comment (indancisive) [inline src/Inputs/Slider.tsx:18]: I think some of your variables can be more clear
at a glance I have no clue what this means or what these values are for

File: src/Inputs/Slider.tsx (hunk 12 above)
Comment (sahandfarahani) [reply src/Inputs/Slider.tsx:52]: yes these are Initial Position Values of the knobs

File: src/Containers/Popup.tsx [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    height,
+    popup,
+}): React.ReactElement => {
+    return (
+        <Container popup={popup} left={left} top={top}>
+            <Rectangle width={width} height={height}>
+                {children}
+            </Rectangle>
+            <Triangle>
+                <InnerTriangle> </InnerTriangle>
```
Comment (aslamm29) [inline src/Containers/Popup.tsx:27]: might want to change it to a self closed tag if not being enclosed with child/data

File: stories/Inputs/Slider.stories.js [hunk 17]
```diff
@@ -1,18 +1,91 @@
...
+        'with popup',
+        mockElement(
+            ([state, setState]) => (
+                <Slider
+                    min={20}
+                    max={50}
+                    padding="50px"
+                    hasPopup={boolean('hasPopup', true)}
+                    onChange={target => {
+                        console.log(target);
```
Comment (aslamm29) [inline stories/Inputs/Slider.stories.js:55]: these are stories but perhaps this log may not be needed in this version here as there is a setState

File: stories/Inputs/Slider.stories.js [hunk 18]
```diff
@@ -1,18 +1,91 @@
...
+            ([state, setState]) => (
+                <Slider
+                    hasTwoKnobs={boolean('twoinputs', true)}
+                    hasRail={boolean('hasRail', true)}
+                    hasPopup={boolean('hasPopup', true)}
+                    min={100}
+                    max={300}
+                    step={50}
+                    onChange={target => {
+                        console.log(target);
```
Comment (aslamm29) [inline stories/Inputs/Slider.stories.js:76]: this one too

File: stories/Containers/Popup.stories.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+    'with nothing',
+    () => (
+        <Popup
+            left={0}
+            top={0}
+            width={0}
+            height={0}
+            popup={true}
+            content="Content"
+        ></Popup>
```
Comment (aslamm29) [inline stories/Containers/Popup.stories.js:15]: marking this tag here

File: stories/Inputs/Slider.stories.js [hunk 20]
```diff
@@ -1,18 +1,91 @@
...
+        mockElement(
+            ([state, setState]) => (
+                <Slider
+                    min={1}
+                    max={100}
+                    onChange={target => {
+                        setState(target);
+                    }}
+                    values={state}
+                ></Slider>
```
Comment (aslamm29) [inline stories/Inputs/Slider.stories.js:20]: marking these and the ones below

File: src/Inputs/Slider.tsx [hunk 21]
```diff
@@ -1,79 +1,384 @@
...
+         top: -21px;
+`
+            : 'top: -7px; '}
+
+    cursor: grab;
+    ${({ theme }): string => `
+            background: ${theme.colors.primary};
+        `}
+    &:active {
+        border: solid 2px #96dbfa;
```
Comment (ryanarine) [inline src/Inputs/Slider.tsx:349]: add `box-sizing: border-box` so the border doesn't change the position of the knob

File: src/Inputs/Slider.tsx [hunk 22]
```diff
@@ -1,79 +1,384 @@
...
+    );
+
+    const [leftKnobPosition, setLeftKnobPosition] = useState(values.leftValue);
+
+    const [isRightKnobDragging, setIsRightKnobDragging] = useState(false);
+    const [isLeftKnobDragging, setIsLeftKnobDragging] = useState(false);
+
+    const maxAndMinDifference = useMemo((): number => {
+        return max - min;
+    }, [max, min]);
```
Comment (ryanarine) [inline src/Inputs/Slider.tsx:78]: Does this need to be in a use Memo? why not just `const difference = max - min`

File: src/Inputs/Slider.tsx [hunk 23]
```diff
@@ -1,79 +1,384 @@
...
+            ) {
+                setRightKnobPosition(newPosition);
+            } else if (isLeftKnobDragging && newPosition < rightKnobPosition) {
+                setLeftKnobPosition(newPosition);
+            }
+        }
+    };
+
+    const onMouseUp = (): void => {
+        document.removeEventListener('mouseup', onMouseUp);
```
Comment (ryanarine) [inline src/Inputs/Slider.tsx:131]: should probably add setIsRightKnobDragging(false) and setIsLeftKnobDragging(false) here

File: src/Inputs/Slider.tsx (hunk 22 above)
Comment (ralph-dev) [reply src/Inputs/Slider.tsx:78]: @ryanarine You don't want to re-calculate the difference everytime one of the other props changes though.

File: src/Inputs/Slider.tsx [hunk 24]
```diff
@@ -1,79 +1,384 @@
...
+                setIsRightKnobDragging(false);
+                setIsLeftKnobDragging(true);
+                // in Case user clicks on the Bar
+                if (
+                    e.target === barRef.current ||
+                    e.target === selectedBarRef.current
+                ) {
+                    setLeftKnobPosition(roundToSteps(clickedOn));
+                }
+            }
```
Comment (ryanarine) [inline src/Inputs/Slider.tsx:169]: You can make this code DRY by doing some thing like
 `const isOnRightSide = clickedOn >= ... || !hasTwoKnobs`
`setIsRightKnobDragging(isOnRightSide)`;
`setIsLeftKnobDragging(!isOnRightSide);`
`if (e.target ...) {isOnRightSide ? setRightKnobPosition(...) : setLeftKnob...}`

File: src/Inputs/Slider.tsx (hunk 22 above)
Comment (ryanarine) [reply src/Inputs/Slider.tsx:78]: @ralph-dev The overhead of useMemo probably outweighs the time it takes to subtract two numbers

Comment (ralph-dev) [reply src/Inputs/Slider.tsx:78]: I agree but in that case just have the function without useMemo at all? You never want things hanging in the component because the calculation will execute almost ever re-render essentially because it's a funtional component.

File: src/Inputs/Slider.tsx (hunk 24 above)
Comment (sahandfarahani) [reply src/Inputs/Slider.tsx:169]: @ryanarine its excellent...  but there is a problem on the last part, I will have to make it an assignmet like const something=isOnRightSide ? setRightKnobPosition(...) : setLeftKnob.... but I guess thats not very semantic right?

**Comment by sahandfarahani** [comment]
> The knobs will never drag the first time you click on them. This is because setIsRightKnobDragging(true) is asynchronous, so the mousemove listener will always be using the old value.
> 
Oh!... I didn't notice this... thank you for pointing it!

**Comment by sahandfarahani** [comment]
> The knobs will never drag the first time you click on them. This is because setIsRightKnobDragging(true) is asynchronous, so the mousemove listener will always be using the old value.
> 
> I like how the right knob both serves as the right knob and the single knob when there is only one knob, rather than making a third knob.

Thank you so much ! I didn't notice that!

File: src/Inputs/Slider.tsx (hunk 24 above)
Comment (ryanarine) [reply src/Inputs/Slider.tsx:169]: I would just remove the ternary operator and replace it with the if-else equivalent

File: src/Inputs/Slider.tsx (hunk 22 above)
Comment (ryanarine) [reply src/Inputs/Slider.tsx:78]: Regardless if its a function or not, wouldn't the calculation execute every re-render? (with useMemo the calculation wouldn't execute every re-render but the useMemo itself would)

Does it matter that this is a functional component? if it was class component , the calculation would have to be in the render function(I think) which would execute every re-render (unless you have shouldComponentUpdate)

Comment (ralph-dev) [reply src/Inputs/Slider.tsx:78]: @ryanarine Not really because if it was just a function then on a re-render it would just be a function creation not a calculation. UseMemo only helps when someone calls this function to not do the calculation again unless 2 variables changes

File: src/Inputs/Slider.tsx [hunk 25]
```diff
@@ -153,9 +153,12 @@ export const Slider: React.FunctionComponent<SliderProps> = ({
             }
         }
     };
+
     useEffect((): void => {
-        document.addEventListener('mousemove', onMouseMove as any);
-        document.addEventListener('mouseup', onMouseUp);
+        if (isRightKnobDragging === true || isLeftKnobDragging === true) {
```
Comment (ryanarine) [inline src/Inputs/Slider.tsx:158]: can do if (isRightKnobDragging || isLeftKnobDragging)

File: src/Containers/Popup.tsx [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+    width: 0;
+    height: 0;
+    border-left: 21px solid transparent;
+    border-right: 0px solid transparent;
+    border-top: 24px solid #fff;
+`;
+
+const Container = styled.div<PopupProps>`
+    width: 100px;
+    height: 100px;
```
Comment (ryanarine) [inline src/Containers/Popup.tsx:66]: Does this need to be here? I removed it and it fixes the overflow problem when the right knob popup is at the end.

**Comment by ralph-dev** [comment]
@indancisive I want you to give the final approval. Look over the code carefully.

File: src/Containers/Popup.tsx [hunk 27]
```diff
@@ -0,0 +1,71 @@
+import React from 'react';
+import styled from 'styled-components';
+import { LabelLayoutProps } from '@Layouts';
+
+export interface PopupProps extends LabelLayoutProps {
+    left?: number;
+    top?: number;
+    width?: number;
+    height?: number;
+    popup?: boolean;
```
Comment (indancisive) [inline src/Containers/Popup.tsx:10]: does this do 'popup' prop do anything?

File: src/Inputs/Input.tsx [hunk 28]
```diff
@@ -1,12 +1,7 @@
...
-import {
-    LabelLayout,
-    LabelLayoutProps,
-    InputFragment,
-    InputFragmentProps,
-} from '../Fragments';
+import { LabelLayout, LabelLayoutProps, InputFragment } from '@Layouts';
 
-export interface InputProps extends LabelLayoutProps, InputFragmentProps {
+export interface InputProps extends LabelLayoutProps {
```
Comment (indancisive) [inline src/Inputs/Input.tsx:4]: why did you remove inputfragmentprops ?

File: src/Inputs/Slider.tsx [hunk 29]
```diff
@@ -1,79 +1,379 @@
...
+        </LabelLayout>
+    );
+};
+
+const SliderBoard = styled.div<SliderProps>`
         width: 100%;
-        height: 8.4px;
-        cursor: pointer;
+        height: 4px;
         animate: 0.2s;
```
Comment (indancisive) [inline src/Inputs/Slider.tsx:253]: is this correct, i don't think animate is a valid attribute?

File: src/Inputs/Slider.tsx [hunk 30]
```diff
@@ -1,79 +1,379 @@
...
+    left: number;
+    right: number;
+    disabled: boolean;
+    hasRail: boolean;
+}
+
+const SliderBoardSelected = styled.div<SelectedBarProps>`
+    top: 1px;
+    height: 4px;
+    animate: 0.2s;
```
Comment (indancisive) [inline src/Inputs/Slider.tsx:280]: here as well, I opened up your branch and it's giving me unknown property, can you just check to make sure this is valid?
image

File: src/Inputs/Slider.tsx [hunk 31]
```diff
@@ -1,79 +1,379 @@
...
+         top: -21px;
+`
+            : 'top: -7px; '}
+
+    cursor: grab;
+    ${({ theme }): string => `
+            background: ${theme.colors.primary};
+        `}
+    &:active {
+        box-sizing: border-box
```
Comment (indancisive) [inline src/Inputs/Slider.tsx:342]: missing semi-colon

File: src/Inputs/Slider.tsx [hunk 32]
```diff
@@ -1,79 +1,379 @@
...
+            ? `
+        cursor: not-allowed;
+        opacity: 0.6;
+    `
+            : ' '}
+    ${({ theme }): string => `
+            background: ${theme.colors.primary};
+        `}
+        &:active {
+        box-sizing: border-box
```
Comment (indancisive) [inline src/Inputs/Slider.tsx:374]: missing semi-colon

File: src/Containers/Popup.tsx [hunk 33]
```diff
@@ -0,0 +1,71 @@
...
+    children,
+    left,
+    top,
+    width,
+    height,
+    popup,
+    ...props
+}): React.ReactElement => {
+    return (
+        <Container popup={popup} left={left} top={top}>
```
Comment (indancisive) [inline src/Containers/Popup.tsx:23]: I don't think pop-up is a standard div attribute can u clarify what this does as it doesn't look like it does anything (I tried passing in false in your story and it just changed nothing)

File: src/Containers/Popup.tsx (hunk 27 above)
Comment (sahandfarahani) [reply src/Containers/Popup.tsx:10]: It's a boolean saying whether component should have popup or not

File: src/Inputs/Input.tsx (hunk 28 above)
Comment (sahandfarahani) [reply src/Inputs/Input.tsx:4]: That's strange... I didnt do anything with input.tsx

File: src/Containers/Popup.tsx (hunk 33 above)
Comment (sahandfarahani) [reply src/Containers/Popup.tsx:23]: Oh it's in popup.tsx
Due to the changes I made to slider.tsx this is useless now thank thank you for pointing it out
